### PR TITLE
Fix evident lack of memory allocations handling in topology-linux.c

### PR DESCRIPTION
**Pull request by kutelev**
Due to not checking return values of some functions hwloc crashes in some cases when memory allocation (malloc) fails. To prevent such crashes there few checks have been added.

### CODE REVIEW

File: hwloc/topology-linux.c [hunk 1]
```diff
@@ -869,6 +871,8 @@ hwloc_linux_find_kernel_nr_cpus(hwloc_topology_t topology)
   fd = open("/sys/devices/system/cpu/possible", O_RDONLY); /* binding only supported in real fsroot, no need for data->root_fd */
   if (fd >= 0) {
     hwloc_bitmap_t possible_bitmap = hwloc_bitmap_alloc_full();
+    if (!possible_bitmap)
+      return -1;
```
Comment (sthibaul) [inline hwloc/topology-linux.c]: fd needs to be closed before returning

File: hwloc/topology-linux.c [hunk 2]
```diff
@@ -1851,9 +1875,11 @@ hwloc_linux_get_area_membind(hwloc_topology_t topology, const void *addr, size_t
   int pagesize = hwloc_getpagesize();
   char *tmpaddr;
   int err;
-  unsigned i;
+  int i;
```
Comment (sthibaul) [inline hwloc/topology-linux.c]: This seems useless?

**Comment by bgoglin** [comment]
Just for fun: Did you actually observe those exact crashes? There are other places where we don't check allocations but I am too lazy to fix all of them...

**Comment by kutelev** [comment]
Yes, we really observed these crashes and to be honest only places where it was really crashing have been fixed.

File: hwloc/topology-linux.c (hunk 2 above)
Comment (kutelev) [inline hwloc/topology-linux.c]: Reverted.

File: hwloc/topology-linux.c (hunk 1 above)
Comment (kutelev) [inline hwloc/topology-linux.c]: Thanks for pointing an issue. Now should be fixed.

**Comment by bgoglin** [comment]
I'd like to reproduce that to make sure we don't need other more important allocation checks later in the discovery process in topology.c. Did you use a fuzzer to generate false allocation failures? Or did you really fill your entire address space to make malloc fail?

**Comment by bgoglin** [comment]
Please squash your commits into a single one (and push -f) so that the review process doesn't appear in the git history. And please revert the other int/unsigned changes, those are unrelated to malloc failures.

**Comment by kutelev** [comment]
hwloc_linux_find_kernel_max_numnodes now returns -1 on error that is why max_os_index type was changed from unsigned to int. Any suggestion how hwloc_linux_find_kernel_max_numnodes can report an error without returning -1? Thanks in advance.

**Comment by bgoglin** [comment]
Just testing whether max_os_index == (unsigned)-1 should work fine without changing the non-error case (that piece of code is hard to test for regressions).

**Comment by kutelev** [comment]
Regarding to the reproducibility in tests allocations are failed synthetically using simple library which is preloaded using LD_PRELOAD and DYLD_INSERT_LIBRARIES. Allocations are failed randomly using different strategies to cover as much as possible cases inclusing completelly abnormal. But in any case due to strict requirements in our case all situations must be handled. Application shall never crash.

**Comment by bgoglin** [comment]
OK that's what I called a fuzzer. I am surprised that no other problem appeared outside of this file but I am not going to complain about it.

**Comment by kutelev** [comment]
Such crazy test exists only for Linux and Mac OS X now. Windows in plans. Maybe other issues will be found there later.

**Comment by kutelev** [comment]
Single commit has been kept. All int/unsigned changes have been reverted.

**Comment by bgoglin** [comment]
Thanks, I'll merge this to master and v1.11 once the testing infrastructure is fixed (hopefully before the end of the week).
If you can share your way to perform this malloc fuzzing testing and maybe even some way to automatize it, that'd be nice.

**Comment by kutelev** [comment]
Thank You. About sharing the approach I will be honest and will not promise that I will provide the working solution immediately. In fact all I can share is library which fails allocations when preloaded. Still not sure that there is high value in this library, it contains ~200 lines of quite simple code. If you still interested in it I can prepare an example application which is tested using such approach, but it will take some time in any case.

**Comment by mofr** [comment]
Why using `(unsigned)-1` was considered as a better approach?

**Comment by kutelev** [comment]
After digging deeper we found (as was predicted) other places where hwloc fails. For example hwloc_topology_init crashes in some cases. As there are more failures that we may want to see simple project which can demonstrate the problem has been created. You can find it there: https://github.com/kutelev/hwlocoverthrower/tree/master/hwlocoverthrower. In the .travis.yml file you can sequence of command using which problem can be reproduced. It uses a simple library (overthrower) which was mentioned earlier. Test application itself is extremely simple, all it does is:
```c
for (i = 0; i < 1000; ++i) {
    if (hwloc_topology_init(&topology) != 0)
        continue;
    hwloc_topology_destroy(topology);
}
```
When overthrower is not preloaded this code finishes correctly, but when is overthrower preloaded this code crashes with probability which is close to 100%. Build on Travis CI can be found there: https://travis-ci.org/kutelev/hwlocoverthrower. Now I shared all I can.

**Comment by bgoglin** [comment]
hwloc_topology_init() would be easy to fix. If you had load() between init() and destroy(), it would be a lot harder.

I can't find the code for liboverthrower. But we could use something else like libfuzzer anyway.

**Comment by kutelev** [comment]
In real life we have hwloc_topology_load after hwloc_topology_init. Regarding to the liboverthrower it is stored in separate a repository.

Full list of commands which can be executed to retrieve and all necessary dependencies:
```bash
git clone https://github.com/kutelev/hwlocoverthrower.git
cd hwlocoverthrower
git clone --depth 1 https://github.com/open-mpi/hwloc.git
git clone --depth 1 https://github.com/kutelev/overthrower.git
cd hwloc
./autogen.sh
./configure --disable-libxml2 --enable-static --disable-shared --disable-libudev --disable-pci
make
cd ../overthrower
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=1 .
cmake --build . --config Release
cd ../hwlocoverthrower
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=1 .
cmake --build . --config Release
cd ..
python ./tests/test.py
```

**Comment by bgoglin** [comment]
Replaced by #248

**Pull request closed without merging**